### ISSUE TITLE
RTDETRv2

### ISSUE THREAD

**Issue by lyuwenyu** [lyuwenyu/RT-DETR#179]
https://arxiv.org/abs/2407.17140

---

I will release the rtdetrv2 soon, which will provide a better deployment experience and higher accuracy.

**You can start this repo to follow news**

---

The AP of **rtdetrv2_r18vd** on the COCO-val2017 dataset is increased by 1.4 and maintains the same inference speed compared with *RTDETRv1*. In addition, in order to make rtdetr everywhere, we provide **rtdetrv2_r18vd_dsp** that utilizes `discrete deformable attention` instead of `deformable attention` with negligible differences on AP<sub>50</sub><sup>val</sup> compared with rtdetrv2_r18vd.

<details open>
<summary> Figure </summary>

<div align="center">
<img width="500" alt="image" src="https://github.com/user-attachments/assets/437877e9-1d4f-4d30-85e8-aafacfa0ec56">
</div>

</details>

**Comment by CatfishW** [comment]
v2有哪些结构上的变化呀ovo

**Comment by huihui308** [comment]
Do you have smaller model which can train on 3080ti?

**Comment by Ryanshuai** [comment]
求大佬更新pytorch v2！

**Comment by leoxxxxxD** [comment]
大佬 年前可能开源吗

**Comment by leoxxxxxD** [comment]
@lyuwenyu 大佬 v2年前可能开源吗

**Comment by lyuwenyu** [comment]
> @lyuwenyu 大佬 v2年前可能开源吗

@leoxxxxxD  不好意思 客观原因 年前放不了了

**Comment by BUG1989** [comment]
@lyuwenyu 大佬，来催更了。之前的预训练模型里面的 add op 有 infinity 数据。NPU 部署不太 make sense 呀

**Comment by BUG1989** [comment]
> @lyuwenyu 大佬，来催更了。之前的预训练模型里面的 add op 有 infinity 数据。模型量化会有问题 ![image](https://private-user-images.githubusercontent.com/13964381/305873754-1c58f394-47c8-4620-ba04-0f912b503d91.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1NjU2MjQsIm5iZiI6MTcwODU2NTMyNCwicGF0aCI6Ii8xMzk2NDM4MS8zMDU4NzM3NTQtMWM1OGYzOTQtNDdjOC00NjIwLWJhMDQtMGY5MTJiNTAzZDkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIyVDAxMjg0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAyZmNhMTEyZGUzZWRlMmFhZGQwZDI4MTYwZjNlZjA4Y2JlN2E0MGRhZTQxYTM1ZmI0MTllZjRkZDI1MmVhOGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Gi8ctq-ksbJpUq9cW8HoLxo3s1eKMUk0M1LWkyW7PXw)

这个问题我们已经解决了

**Comment by SoulProficiency** [comment]
RTDETR可以转rknn吗？

**Comment by SoulProficiency** [comment]
> 我的问题我们已经解决了，在我司的芯片上运行效果挺好。 ![b7ace8de6dfabb44cce40f5677638a7](https://private-user-images.githubusercontent.com/13964381/306823318-98bacff2-cb2d-40d5-a16f-8d587d67d850.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5NDE2NzgsIm5iZiI6MTcwODk0MTM3OCwicGF0aCI6Ii8xMzk2NDM4MS8zMDY4MjMzMTgtOThiYWNmZjItY2IyZC00MGQ1LWExNmYtOGQ1ODdkNjdkODUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI2VDA5NTYxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE2NDM3ZjU2MmRkZmUxMDRkMGY4MmUxODY2YTc5ODRiY2UzYzY1YzI4ZWFhYTk2YzJlNmM5NDcxZDYxNzU3YzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9wEde4E8bNzGvZtUgAcYMKnArTFVr-94sFeyv-vod9I)

老哥，你们的芯片是rk的吗？

**Comment by zhouzq-thu** [comment]
waiting

**Comment by Peterande** [comment]
您好，请问v2有投稿意向吗？能得知大概会在什么时间范围内放出嘛？
我也有意向将RT-DETR这个IP做下去，像YOLO一样形成V2,V3....并且也取得了一些成果，
但似乎在您的官方v2发布前我无法擅自进行这种延续性的命名qwq

**Comment by lyuwenyu** [comment]
> 您好，请问v2有投稿意向吗？能得知大概会在什么时间范围内放出嘛？ 我也有意向将RT-DETR这个IP做下去，像YOLO一样形成V2,V3....并且也取得了一些成果， 但似乎在您的官方v2发布前我无法擅自进行这种延续性的命名qwq

---

@Peterande 非常抱歉本`issue`名字给你带来的困扰

---

另外也非常高兴社区有`RTDETR`方向的优化，非常期待成果。( 这也是我们开源的目的之一 )
1. `RTDETRv2`是我们刚发布`v1`之后的产物，是我们对`RTDETR`的优化版本（主要是训练策略and部署工程方面的优化，模型结构和`v1`是一致的，可能叫`1.5`更合适）。**所以从创新的的角度，应该不会和你有冲突**
2. `RTDETRv2`模型本来是有开源的计划，但是由于缺少人手的问题目前未能开源，也未形成技术报告（模型限内部使用，但是不排除之后会直接开源代码）
3. 鉴于本`discussion (comming soon)`已经存在的事实，**建议你直接用`v3`，但是不需要和`v2`进行比较**

如果有疑问和其他想法 欢迎继续讨论

**Comment by Peterande** [comment]
> > 您好，请问v2有投稿意向吗？能得知大概会在什么时间范围内放出嘛？ 我也有意向将RT-DETR这个IP做下去，像YOLO一样形成V2,V3....并且也取得了一些成果， 但似乎在您的官方v2发布前我无法擅自进行这种延续性的命名qwq
> 
> @Peterande 非常抱歉本`issue`名字给你带来的困扰
> 
> 另外也非常高兴社区有`RTDETR`方向的优化，非常期待成果。( 这也是我们开源的目的之一 )
> 
> 1. `RTDETRv2`是我们刚发布`v1`之后的产物，是我们对`RTDETR`的优化版本（主要是训练策略and部署方面的优化，模型结构和`v1`是一致的，可能叫`v1.5`更合适）。**所以从创新的的角度，应该不会和你有冲突**
> 2. `RTDETRv2`模型本来是有开源的计划，但是由于缺少人手的问题目前未能开源，也未形成技术报告/论文（模型限内部使用，但是不排除之后会直接开源代码）
> 3. 鉴于本`discussion (comming soon)`已经存在的事实，**建议你直接用`v3`，但是论文不需要和`v2`进行比较**
> 
> 如果有疑问和其他想法 欢迎继续讨论

非常感谢您的解答和分享！我已经明白了RTDETRv2的情况，也感谢您团队对社区的贡献和开源精神。

**Comment by lyuwenyu** [comment]
https://arxiv.org/abs/2407.17140

**Comment by Ryanshuai** [comment]
> https://arxiv.org/abs/2407.17140

感谢感谢，现在就卡在 deformable attention，好多板子支持的不好。
感谢！

**Comment by defrag-bambino** [comment]
源代码什么时候到？

> https://arxiv.org/abs/2407.17140

**Comment by imyhxy** [comment]
你好，我使用`rtdetv2_r18vd_120e_coco.yaml`训练，最终结果为47.6，与论文中的48.1存在一定差距，请问有可能是什么原因。

这是训练命令：
```bash
export CUDA_VISIBLE_DEVICES=0,1,2,3

python -m torch.distributed.run \
  --master_port=9909 \
  --nproc_per_node=4 \
  tools/train.py \
  --config configs/rtdetrv2/rtdetrv2_r18vd_120e_coco.yml \
  --use-amp \
  --seed 0
```

这是训练记录：
[log.txt](https://github.com/user-attachments/files/17286901/log.txt)

训练结果截图：
![image](https://github.com/user-attachments/assets/9c91cada-3a67-4b20-81d8-b645e02e1a80)

**Comment by lyuwenyu** [comment]
@imyhxy    使用最新的torch版本试一下, 这是48.1的log
https://github.com/lyuwenyu/storage/releases/download/v0.2/log_rtdetrv2_r18vd_120e_coco_rerun_48.1.txt

**Comment by lyuwenyu** [comment]
> Hi How can I create a Confusion Matrix for my val/test set?

This repo has no relevant tools. You need to find third-party tools.  @babakbch

**Comment by imyhxy** [comment]
@lyuwenyu 我使用了torch 2.4.1来训练的，我也对比了提供日志中的超参数，和我训练的是一致的。
请问你使用了的pytorch，opencv，CUDA和cudnn版本是多少。

**Comment by lyuwenyu** [comment]
这些都没有记录 日志上传的时间是在8月份时候  torch应该是用的2.4吧。 不好意思 时间有点久了 机器已经释放了。 @imyhxy

**Comment by flytocc** [comment]
> 你好，我使用`rtdetv2_r18vd_120e_coco.yaml`训练，最终结果为47.6，与论文中的48.1存在一定差距，请问有可能是什么原因。

我把rtdetrv2移植到mmdet上试了一下[repo](https://github.com/flytocc/mmdetection/tree/rtdetr-v2)，精度47.5 (torch 2.2.1)

**Comment by lyuwenyu** [comment]
这是完整的log可以参考下，另外建议用最新的torch和torchvision。 @flytocc  
https://github.com/lyuwenyu/storage/releases/download/v0.2/log_rtdetrv2_r18vd_120e_coco_rerun_48.1.txt

**Comment by valldynsyahdanuarta** [comment]
when i train this model into custom data it generate log files and it just contain the train loss, so my question is how i can add the validation loss when training and plot it togethe?r.

**Comment by zoldaten** [comment]
is RTDETRv2 really bit YOLO?

i have compared speeds on raspberry pi 4b and got results as follows:
RT-DETRv2:
RT-DETRv2-S  pt version - 6-7 sec
RT-DETRv2-S onnx version - 3-4 sec
RT-DETRv2-S onnx simlify - 3 sec
https://www.youtube.com/watch?v=qYkq9wydcKY

YOLO:
yolo11n.pt 640x640 - 1,4 sec  
 yolo11n.onnx imgsz=dynamic int8 - 726ms
yolo11n.onnx imgsz=256 int8 - 333ms
https://www.youtube.com/watch?v=0rnizyI0ty4

**Comment by zhang123-lf** [comment]
> > > 我的问题我们已经解决了，在我司的芯片上运行效果很好。![b7ace8de6dfabb44cce40f5677638a7](https://private-user-images.githubusercontent.com/13964381/306823318-98bacff2-cb2d-40d5-a16f-8d587d67d850.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5NDE2NzgsIm5iZiI6MTcwODk0MTM3OCwicGF0aCI6Ii8xMzk2NDM4MS8zMDY4MjMzMTgtOThiYWNmZjItY2IyZC00MGQ1LWExNmYtOGQ1ODdkNjdkODUwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI2VDA5NTYxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE2NDM3ZjU2MmRkZmUxMDRkMGY4MmUxODY2YTc5ODRiY2UzYzY1YzI4ZWFhYTk2YzJlNmM5NDcxZDYxNzU3YzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9wEde4E8bNzGvZtUgAcYMKnArTFVr-94sFeyv-vod9I)
> > 
> > 
> > 老哥，你们的芯片是rk的吗？
> 
> 爱芯元智，AX650N

大佬，请问rtdetrv2推理的帧率能达到多少呢
